File: rust/benchmarks/a2c.rs
use criterion::{criterion_group, criterion_main, BatchSize, Criterion, PlottingBackend};
use tch::{nn::VarStore, Device};

const MAX_ITERATION_COUNT: usize = 500_000;
const EVAL_INTERVAL: usize = 500;

const TARGET_AVG_RETURN: f64 = 475.0;

fn a2c_cpu(c: &mut Criterion) {
    use utils::{a2c::A2C, env};

    c.bench_function("A2C/CPU", |b| {
        b.iter_batched(
            || {
                let env = env::MyVeryOwnStick::new();

                let vs = VarStore::new(Device::Cpu);

                A2C::builder()
                    .vs(vs)
                    .env(env)
                    .fc_layers(&[100, 100])
                    .gamma(0.99)
                    .learning_rate(1e-3)
                    .sync_interval(100)
                    .build()
            },
            |mut a2c| {
                for i in 0..(MAX_ITERATION_COUNT / EVAL_INTERVAL) {
                    a2c.train(EVAL_INTERVAL);

                    let avg_return = a2c.evaluate_avg_return();

                    if avg_return >= TARGET_AVG_RETURN {
                        break;
                    }

                    assert!(
                        (i + 1) * EVAL_INTERVAL != MAX_ITERATION_COUNT,
                        "Failed to reach target average return"
                    );
                }
            },
            BatchSize::SmallInput,
        );
    });
}

criterion_group! {
    name = benches;
    config = Criterion::default()
        .sample_size(200)
        .plotting_backend(PlottingBackend::None);
    targets = a2c_cpu
}
criterion_main!(benches);

File: python/benchmarks/a2c.py
import statistics
import time

from utils.a2c import A2C
from utils.my_very_own_stick_env import MyVeryOnwStickEnv

SAMPLE_SIZE = 50
MAX_ITERATION_COUNT = 500_000
EVAL_INTERVAL = 500

TARGET_AVG_RETURN = 475.0

durations = []

for _ in range(SAMPLE_SIZE):
    a2c = A2C(
        env=MyVeryOnwStickEnv(),
        learning_rate=1e-3,
        sync_interval=100,
        gamma=0.99,
        fc_layer_params=(100, 100)
    )

    start_time = time.perf_counter()

    for i in range(MAX_ITERATION_COUNT // EVAL_INTERVAL):
        a2c.train(EVAL_INTERVAL)
        avg_return = a2c.evaluate_avg_return()

        if avg_return >= TARGET_AVG_RETURN:
            break

        if (i + 1) * EVAL_INTERVAL == MAX_ITERATION_COUNT:
            raise RuntimeError("Failed to reach target average return")

    durations.append(time.perf_counter() - start_time)

print(f"Median duration: {statistics.median(durations)}s")
